geode-core/src/main/java/org/apache/geode/internal/net/SCAdvancedSocketCreator.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.internal.net;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+
+import javax.net.SocketFactory;
+
+import org.apache.geode.GemFireConfigException;
+import org.apache.geode.SystemConnectException;
+import org.apache.geode.distributed.internal.tcpserver.AdvancedSocketCreatorImpl;
+import org.apache.geode.distributed.internal.tcpserver.ConnectionWatcher;
+import org.apache.geode.distributed.internal.tcpserver.HostAndPort;
+
+class SCAdvancedSocketCreator extends AdvancedSocketCreatorImpl {
+  final SocketCreator coreSocketCreator;
+
+  protected SCAdvancedSocketCreator(SocketCreator socketCreator) {
+    super(socketCreator);
+    coreSocketCreator = socketCreator;
+  }
+
+  @Override
+  public void handshakeIfSocketIsSSL(Socket socket, int timeout) throws IOException {
+    coreSocketCreator.handshakeIfSocketIsSSL(socket, timeout);
+  }
+
+  @Override
+  public Socket connect(HostAndPort addr, int timeout,
+      ConnectionWatcher optionalWatcher, boolean allowClientSocketFactory, int socketBufferSize,
+      boolean useSSL) throws IOException {
+
+    coreSocketCreator.printConfig();
+
+    if (!useSSL) {
+      return super.connect(addr, timeout, optionalWatcher, allowClientSocketFactory,
+          socketBufferSize,
+          useSSL);
+    }
+
+    // create an SSL connection
+
+    Socket socket;
+    InetSocketAddress sockaddr = addr.getSocketInetAddress();
+    if (sockaddr.getAddress() == null) {
+      InetAddress address = InetAddress.getByName(sockaddr.getHostString());
```

Comment:
…then, if address resolution failed, we try `InetAddress.getByName()` again. With the exact same hostname.

Two things:

1. I don't see how this second address resolution will ever work if the first one failed (barring some rare race with DNS data)
2. In the case where we enter the conditional at line 61, we end up with `sockaddr` referring to an `InetSocketAddress` that has a `null` address and _no_ hostname at all. Wheres if we do not enter the conditional `sockaddr` refers to an `InetSocketAddress` that has both hostname and port.

Should we remove this conditional entirely?

Reply:
Ah now I see that line 61 throws an `UnknownHostException` if address resolution fails. This, while quirky, is reasonable. Inasmuch as this logic has been in the product a long time, I'm ok with leaving it as-is.

Reply:
I had to do some recollection on this one Bill.
HostAndPort.getSocketInetAddress() is required to resolve hostnames if possible.  It's supposed to hold onto hostname:port pairs and leave the name unresolved until this method is called.
Yes, the second resolution will fail 100% of the time, but it will give us the expected UnknownHostException with the correct text.  Without that lookup there was at least one test failure that asserted that it should get this exception when attempting to use an invalid host name.